src/fragment-generation-utils.js

```diff
@@ -0,0 +1,64 @@
+/**
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as fragments from './text-fragment-utils.js';
+
+/**
+ * Enum indicating the success, or failure reason, of generateFragment.
+ */
+export const GenerateFragmentStatus = {
+  SUCCESS: 0,            // A fragment was generated.
+  INVALID_SELECTION: 1,  // The selection provided could not be used.
+  AMBIGUOUS: 2  // No unique fragment could be identified for this selection.
+}
+
+/**
+ * @typedef {Object} GenerateFragmentResult
+ * @property {GenerateFragmentStatus} status
+ * @property {TextFragment} [fragment]
+ */
+
+/**
+ * Attempts to generate a fragment, suitable for formatting and including in a
+ * URL, which will highlight the given selection upon opening.
+ * @param {Selection} selection - a Selection object, the result of
+ *     window.getSelection
+ * @return {GenerateFragmentResult}
+ */
+export const generateFragment = (selection) => {
+  let range;
+  try {
+    range = selection.getRangeAt(0);
```

Comment:
Will there always be only one range?

Reply:
Yup. For historical/compatibility reasons the spec says this method requires an argument and throws an exception if that argument is not 0

Javascript 🙃